File: ios/TabletHID/Models/TouchMouseConfig.swift
import Foundation

enum TouchMode: String, Codable, CaseIterable, Identifiable {
    case touch
    case mouse

    var id: String { rawValue }
}

enum ZoneType: String, Codable, CaseIterable, Identifiable {
    case staticZone
    case dynamic

    var id: String { rawValue }
}

enum ClickBehavior: String, Codable, CaseIterable, Identifiable {
    case momentary
    case latching

    var id: String { rawValue }
}

enum MacroHostDefaults: String, Codable, CaseIterable, Identifiable {
    case windows
    case mac

    var id: String { rawValue }
    var label: String { self == .windows ? "Windows" : "Mac" }
}

struct KeyboardMacroButtonConfig: Codable, Equatable, Identifiable {
    var id = UUID()
    var label: String
    var modifiers: Int
    var keyUsages: [Int]
    var layoutOffsetX: Double = 0
    var layoutOffsetY: Double = 0
    var layoutScaleX: Double = 1
    var layoutScaleY: Double = 1

    enum CodingKeys: String, CodingKey {
        case id, label, modifiers, keyUsages, layoutOffsetX, layoutOffsetY, layoutScaleX, layoutScaleY
    }

    init(
        id: UUID = UUID(),
        label: String,
        modifiers: Int,
        keyUsages: [Int],
        layoutOffsetX: Double = 0,
        layoutOffsetY: Double = 0,
        layoutScaleX: Double = 1,
        layoutScaleY: Double = 1
    ) {
        self.id = id
        self.label = label
        self.modifiers = modifiers
        self.keyUsages = keyUsages
        self.layoutOffsetX = layoutOffsetX
        self.layoutOffsetY = layoutOffsetY
        self.layoutScaleX = layoutScaleX
        self.layoutScaleY = layoutScaleY
    }

    init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        id = try values.decodeIfPresent(UUID.self, forKey: .id) ?? UUID()
        label = try values.decodeIfPresent(String.self, forKey: .label) ?? "Macro"
        modifiers = try values.decodeIfPresent(Int.self, forKey: .modifiers) ?? 0
        keyUsages = try values.decodeIfPresent([Int].self, forKey: .keyUsages) ?? []
        layoutOffsetX = try values.decodeIfPresent(Double.self, forKey: .layoutOffsetX) ?? 0
        layoutOffsetY = try values.decodeIfPresent(Double.self, forKey: .layoutOffsetY) ?? 0
        layoutScaleX = try values.decodeIfPresent(Double.self, forKey: .layoutScaleX) ?? 1
        layoutScaleY = try values.decodeIfPresent(Double.self, forKey: .layoutScaleY) ?? 1
    }
}

enum KeyboardMacroPresets {
    static let modLeftControl = 0x01
    static let modLeftAlt = 0x04
    static let modLeftGUI = 0x08
    static let keyS = 0x16
    static let keyTab = 0x2B

    static let windowsDefaults = [
        KeyboardMacroButtonConfig(label: "Alt+Tab", modifiers: modLeftAlt, keyUsages: [keyTab]),
        KeyboardMacroButtonConfig(label: "Ctrl+S", modifiers: modLeftControl, keyUsages: [keyS])
    ]

    static let macDefaults = [
        KeyboardMacroButtonConfig(label: "Cmd+Tab", modifiers: modLeftGUI, keyUsages: [keyTab]),
        KeyboardMacroButtonConfig(label: "Cmd+S", modifiers: modLeftGUI, keyUsages: [keyS])
    ]

    static func defaults(for host: MacroHostDefaults) -> [KeyboardMacroButtonConfig] {
        host == .mac ? macDefaults : windowsDefaults
    }
}

struct ButtonZoneConfig: Codable, Equatable {
    var enabled = false
    var zoneType = ZoneType.staticZone
    var behavior = ClickBehavior.momentary
    var staticLeft: Double = 0
    var staticTop: Double = 0.75
    var staticRight: Double = 0.25
    var staticBottom: Double = 1
    var dynamicOffsetX: Double = 0.15
    var dynamicOffsetY: Double = 0
    var dynamicRadius: Double = 0.07
}

struct TouchMouseConfig: Codable, Equatable {
    var mode = TouchMode.touch
    var sensitivity = 5
    var leftButton = ButtonZoneConfig(dynamicOffsetX: -0.15)
    var rightButton = ButtonZoneConfig(
        staticLeft: 0.75,
        staticTop: 0.75,
        staticRight: 1,
        staticBottom: 1
    )
    var scrollEnabled = true
    var invertScroll = false
    var sharedDynamicZone = false
    var sharedDynamicOffsetX: Double = 0
    var sharedDynamicOffsetY: Double = 0.18
    var sharedDynamicRadius: Double = 0.08
    var macroHostDefaults = MacroHostDefaults.windows
    var macroButtons: [KeyboardMacroButtonConfig] = []

    init(
        mode: TouchMode = .touch,
        sensitivity: Int = 5,
        leftButton: ButtonZoneConfig = ButtonZoneConfig(dynamicOffsetX: -0.15),
        rightButton: ButtonZoneConfig = ButtonZoneConfig(
            staticLeft: 0.75,
            staticTop: 0.75,
            staticRight: 1,
            staticBottom: 1
        ),
        scrollEnabled: Bool = true,
        invertScroll: Bool = false,
        sharedDynamicZone: Bool = false,
        sharedDynamicOffsetX: Double = 0,
        sharedDynamicOffsetY: Double = 0.18,
        sharedDynamicRadius: Double = 0.08,
        macroHostDefaults: MacroHostDefaults = .windows,
        macroButtons: [KeyboardMacroButtonConfig] = []
    ) {
        self.mode = mode
        self.sensitivity = sensitivity
        self.leftButton = leftButton
        self.rightButton = rightButton
        self.scrollEnabled = scrollEnabled
        self.invertScroll = invertScroll
        self.sharedDynamicZone = sharedDynamicZone
        self.sharedDynamicOffsetX = sharedDynamicOffsetX
        self.sharedDynamicOffsetY = sharedDynamicOffsetY
        self.sharedDynamicRadius = sharedDynamicRadius
        self.macroHostDefaults = macroHostDefaults
        self.macroButtons = macroButtons
    }

    enum CodingKeys: String, CodingKey {
        case mode, sensitivity, leftButton, rightButton, scrollEnabled, invertScroll
        case sharedDynamicZone, sharedDynamicOffsetX, sharedDynamicOffsetY, sharedDynamicRadius
        case macroHostDefaults, macroButtons
    }

    init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        let defaults = TouchMouseConfig()
        mode = try values.decodeIfPresent(TouchMode.self, forKey: .mode) ?? defaults.mode
        sensitivity = try values.decodeIfPresent(Int.self, forKey: .sensitivity) ?? defaults.sensitivity
        leftButton = try values.decodeIfPresent(ButtonZoneConfig.self, forKey: .leftButton) ?? defaults.leftButton
        rightButton = try values.decodeIfPresent(ButtonZoneConfig.self, forKey: .rightButton) ?? defaults.rightButton
        scrollEnabled = try values.decodeIfPresent(Bool.self, forKey: .scrollEnabled) ?? defaults.scrollEnabled
        invertScroll = try values.decodeIfPresent(Bool.self, forKey: .invertScroll) ?? defaults.invertScroll
        sharedDynamicZone = try values.decodeIfPresent(Bool.self, forKey: .sharedDynamicZone) ?? defaults.sharedDynamicZone
        sharedDynamicOffsetX = try values.decodeIfPresent(Double.self, forKey: .sharedDynamicOffsetX) ?? defaults.sharedDynamicOffsetX
        sharedDynamicOffsetY = try values.decodeIfPresent(Double.self, forKey: .sharedDynamicOffsetY) ?? defaults.sharedDynamicOffsetY
        sharedDynamicRadius = try values.decodeIfPresent(Double.self, forKey: .sharedDynamicRadius) ?? defaults.sharedDynamicRadius
        macroHostDefaults = try values.decodeIfPresent(MacroHostDefaults.self, forKey: .macroHostDefaults) ?? defaults.macroHostDefaults
        macroButtons = try values.decodeIfPresent([KeyboardMacroButtonConfig].self, forKey: .macroButtons) ?? defaults.macroButtons
    }

    func normalizedForStorage() -> TouchMouseConfig {
        var config = self
        config.leftButton = config.leftButton.normalizedForStorage()
        config.rightButton = config.rightButton.normalizedForStorage()
        config.sharedDynamicOffsetX = config.sharedDynamicOffsetX.clamped(to: -1...1).snapped(to: 0.05)
        config.sharedDynamicOffsetY = config.sharedDynamicOffsetY.clamped(to: -1...1).snapped(to: 0.05)
        config.sharedDynamicRadius = config.sharedDynamicRadius.clamped(to: 0.03...0.2).snapped(to: 0.01)
        return config
    }

    static func defaultForProfile(_ profile: Profile) -> TouchMouseConfig {
        var config = TouchMouseConfig()
        config.mode = .mouse
        config.sensitivity = profile.key == Profile.accessAdvanced.key ? 3 : 7
        config.leftButton.enabled = true
        config.leftButton.zoneType = .dynamic
        config.leftButton.dynamicOffsetY = 0.2
        config.leftButton.dynamicRadius = 0.1
        config.rightButton.enabled = true
        config.rightButton.zoneType = .dynamic
        config.rightButton.dynamicOffsetX = 0.2
        config.rightButton.dynamicOffsetY = 0.2
        return config
    }
}

private extension ButtonZoneConfig {
    func normalizedForStorage() -> ButtonZoneConfig {
        var button = self
        button.staticLeft = button.staticLeft.clamped(to: 0...1).snapped(to: 0.01)
        button.staticTop = button.staticTop.clamped(to: 0...1).snapped(to: 0.01)
        button.staticRight = button.staticRight.clamped(to: 0...1).snapped(to: 0.01)
        button.staticBottom = button.staticBottom.clamped(to: 0...1).snapped(to: 0.01)
        button.dynamicOffsetX = button.dynamicOffsetX.clamped(to: -1...1).snapped(to: 0.05)
        button.dynamicOffsetY = button.dynamicOffsetY.clamped(to: -1...1).snapped(to: 0.05)
        button.dynamicRadius = button.dynamicRadius.clamped(to: 0.03...0.2).snapped(to: 0.01)
        return button
    }
}

extension Double {
    func clamped(to range: ClosedRange<Double>) -> Double {
        min(max(self, range.lowerBound), range.upperBound)
    }

    func snapped(to step: Double) -> Double {
        guard step > 0 else { return self }
        return (self / step).rounded() * step
    }
}

File: app/src/main/java/com/tablet/hid/model/TouchMouseConfig.kt
package com.tablet.hid.model

enum class TouchMode { TOUCH, MOUSE }
enum class ZoneType { STATIC, DYNAMIC }
enum class ClickBehavior { MOMENTARY, LATCHING }
enum class MouseButton(val bit: Int) { LEFT(1), RIGHT(2), MIDDLE(4) }
enum class MacroHostDefaults { WINDOWS, MAC }

data class KeyboardMacroButtonConfig(
    val label: String,
    val modifiers: Int,
    val keyUsages: List<Int>,
)

object KeyboardMacroPresets {
    const val MOD_LEFT_CONTROL = 0x01
    const val MOD_LEFT_ALT = 0x04
    const val MOD_LEFT_GUI = 0x08
    const val KEY_S = 0x16
    const val KEY_TAB = 0x2B

    val windowsDefaults = listOf(
        KeyboardMacroButtonConfig("Alt+Tab", MOD_LEFT_ALT, listOf(KEY_TAB)),
        KeyboardMacroButtonConfig("Ctrl+S", MOD_LEFT_CONTROL, listOf(KEY_S)),
    )

    val macDefaults = listOf(
        KeyboardMacroButtonConfig("Cmd+Tab", MOD_LEFT_GUI, listOf(KEY_TAB)),
        KeyboardMacroButtonConfig("Cmd+S", MOD_LEFT_GUI, listOf(KEY_S)),
    )

    fun defaultsFor(host: MacroHostDefaults): List<KeyboardMacroButtonConfig> =
        if (host == MacroHostDefaults.MAC) macDefaults else windowsDefaults
}

data class TouchMouseSubRegionConfig(
    val enabled: Boolean = true,
    val zoneType: ZoneType = ZoneType.STATIC,
    // Static zone: fractional screen coordinates in [0, 1]
    val staticLeft: Float = 0f,
    val staticTop: Float = 0f,
    val staticRight: Float = 0f,
    val staticBottom: Float = 0f,
    // Dynamic zone: offset from primary pointer as fraction of min(width, height)
    val dynamicOffsetX: Float = 0f,
    val dynamicOffsetY: Float = 0f,
    val dynamicRadius: Float = 0.07f,
    // Groundwork for future keyboard/mouse-modified sub-region actions.
    val keyboardModifiers: Int = 0,
    val alternateMouseButton: MouseButton? = null,
)

data class ButtonZoneConfig(
    val enabled: Boolean = false,
    val zoneType: ZoneType = ZoneType.STATIC,
    val behavior: ClickBehavior = ClickBehavior.MOMENTARY,
    // Static zone: fractional screen coordinates in [0, 1]
    val staticLeft: Float = 0f,
    val staticTop: Float = 0.75f,
    val staticRight: Float = 0.25f,
    val staticBottom: Float = 1f,
    // Dynamic zone: offset from primary pointer as fraction of min(width, height)
    val dynamicOffsetX: Float = 0.15f,
    val dynamicOffsetY: Float = 0f,
    val dynamicRadius: Float = 0.07f,
    val subRegions: List<TouchMouseSubRegionConfig> = emptyList(),
)

data class TouchMouseConfig(
    val mode: TouchMode = TouchMode.TOUCH,
    val sensitivity: Int = 5,
    val scrollEnabled: Boolean = true,
    val invertScroll: Boolean = false,
    val sharedDynamicZone: Boolean = false,
    val sharedDynamicOffsetX: Float = 0f,
    val sharedDynamicOffsetY: Float = 0.18f,
    val sharedDynamicRadius: Float = 0.08f,
    val leftButton: ButtonZoneConfig = ButtonZoneConfig(
        staticLeft = 0f, staticTop = 0.75f, staticRight = 0.25f, staticBottom = 1f,
        dynamicOffsetX = -0.15f   // left of primary pointer in dynamic mode
    ),
    val rightButton: ButtonZoneConfig = ButtonZoneConfig(
        staticLeft = 0.75f, staticTop = 0.75f, staticRight = 1f, staticBottom = 1f,
        dynamicOffsetX = +0.15f // (default) — right of primary pointer
    ),
    val macroHostDefaults: MacroHostDefaults = MacroHostDefaults.WINDOWS,
    val macroButtons: List<KeyboardMacroButtonConfig> = emptyList(),
)
